src/openrct2-ui/windows/ScenarioSelect.cpp

```diff
@@ -254,433 +205,413 @@ static void WindowScenarioselectInitTabs(WindowBase* w)
-            ft.Add<uint16_t>(scenario->ObjectiveArg2);
-        else
-            ft.Add<money64>(scenario->ObjectiveArg2);
+        const int32_t scenarioItemHeight = GetScenarioListItemSize();
+
+        bool originalShowLockedInformation = _showLockedInformation;
+        _showLockedInformation = false;
+        const ScenarioIndexEntry* selected = nullptr;
+        auto mutableScreenCoords = screenCoords;
+        for (const auto& listItem : _listItems)
+        {
+            switch (listItem.type)
+            {
+                case ListItemType::Heading:
+                    mutableScreenCoords.y -= 18;
+                    break;
+                case ListItemType::Scenario:
+                    mutableScreenCoords.y -= scenarioItemHeight;
+                    if (mutableScreenCoords.y < 0)
+                    {
+                        if (listItem.scenario.is_locked)
+                        {
+                            _showLockedInformation = true;
+                        }
+                        else
+                        {
+                            selected = listItem.scenario.scenario;
+                        }
+                    }
+                    break;
+            }
+            if (mutableScreenCoords.y < 0)
+            {
+                break;
+            }
+        }
+
+        if (highlighted_scenario != selected)
+        {
+            highlighted_scenario = selected;
+            Invalidate();
+        }
+        else if (_showLockedInformation != originalShowLockedInformation)
+        {
+            Invalidate();
+        }
     }
-    screenPos.y += DrawTextWrapped(*dpi, screenPos, 170, STR_OBJECTIVE, ft) + 5;
 
-    // Scenario score
-    if (scenario->Highscore != nullptr)
+    void OnScrollMouseDown(int32_t scrollIndex, const ScreenCoordsXY& screenCoords) override
     {
-        // TODO: Should probably be translatable
-        u8string completedByName = "???";
-        if (!scenario->Highscore->name.empty())
+        const int32_t scenarioItemHeight = GetScenarioListItemSize();
+
+        auto mutableScreenCoords = screenCoords;
+        for (const auto& listItem : _listItems)
         {
-            completedByName = scenario->Highscore->name;
+            switch (listItem.type)
+            {
+                case ListItemType::Heading:
+                    mutableScreenCoords.y -= 18;
+                    break;
+                case ListItemType::Scenario:
+                    mutableScreenCoords.y -= scenarioItemHeight;
+                    if (mutableScreenCoords.y < 0 && !listItem.scenario.is_locked)
+                    {
+                        OpenRCT2::Audio::Play(OpenRCT2::Audio::SoundId::Click1, 0, windowPos.x + (width / 2));
+                        gFirstTimeSaving = true;
+                        _callback(listItem.scenario.scenario->Path);
+                        if (_titleEditor)
+                        {
+                            Close();
```

Comment:
You have to return from this function immediately after calling close, which destroys current window and the list item alongside


```suggestion
                            Close();
                            return;
```